### PR TITLE
[SPARK-14637][SQL] object expressions cleanup

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

Simplify and clean up some object expressions:
1. simplify the logic to handle `propagateNull`
2. add `propagateNull` parameter to `Invoke`
3. simplify the unbox logic in `Invoke`
4. other minor cleanup

TODO: simplify `MapObjects`
## How was this patch tested?

existing tests.

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @marmbrus @yhuai @liancheng

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala [hunk 1]
```diff
@@ -130,60 +126,52 @@ case class Invoke(
...
-      s"boolean ${ev.isNull} = ${ev.value} == null;"
+    val callFunc = if (method.isDefined && method.get.getReturnType.isPrimitive) {
+      s"${obj.value}.$functionName($argString)"
     } else {
-      ev.isNull = obj.isNull
-      ""
+      s"(${ctx.boxedType(javaType)}) ${obj.value}.$functionName($argString)"
     }
 
-    val value = unboxer(s"${obj.value}.$functionName($argString)")
```
Comment (liancheng) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: Don't we need to do unboxing anymore here?

Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/objects.scala]: We do, but I simplified it, see the `callFunc` above.

**Comment by marmbrus** [comment]
Sorry this went stale.  Do we still want to try and get this in?

**Comment by cloud-fan** [comment]
retest this please.

**Comment by cloud-fan** [comment]
retest this please

**Comment by marmbrus** [comment]
Thanks, merging to master and 2.0

**Pull request closed without merging**